crypto/crypto-derive/src/lib.rs

```diff
@@ -229,3 +232,60 @@ pub fn derive_enum_signature(input: TokenStream) -> TokenStream {
         }
     }
 }
+
+#[proc_macro_derive(CryptoHasher, attributes(CryptoHasherSalt))]
+pub fn hasher_dispatch(input: TokenStream) -> TokenStream {
+    let item = parse_macro_input!(input as DeriveInput);
+    let hasher_name = Ident::new(
+        &format!("{}Hasher", &item.ident.to_string()),
+        Span::call_site(),
+    );
+    let snake_name = camel_to_snake(&item.ident.to_string());
+    let static_hasher_name = Ident::new(
+        &format!("{}_HASHER", snake_name.to_uppercase()),
+        Span::call_site(),
+    );
+    let fn_name = get_type_from_attrs(&item.attrs, "CryptoHasherSalt")
+        .unwrap_or_else(|_| syn::LitStr::new(&snake_name, Span::call_site()));
+
+    let out = quote!(
+        #[derive(Clone)]
+        pub struct #hasher_name(libra_crypto::hash::DefaultHasher);
+
+        impl #hasher_name {
+            fn new() -> Self {
+                // space-right-padded, hex-encoded path on 8 chars
+                let mp = format!("{path:8.8}", path=::hex::encode(module_path!().as_bytes()));
```

Comment:
I would suggest hashing (sha3) the full path + name, to stay as consistent as possible and at the same time have a fixed_size salt that will help us even if we never pick cShake. As this will happen at compilation time, there is not performance penalty as well.
If we agree on this, then snake_case and hex encoding are not required.

Reply:
See [below](https://github.com/libra/libra/pull/1511#issuecomment-551860576).